torch/csrc/autograd/edge.h

```diff
@@ -0,0 +1,53 @@
+#pragma once
+
+#include <cstdint>
+#include <functional>
+#include <memory>
+
+#include "torch/csrc/utils/hash.h"
+
+namespace torch {
+namespace autograd {
+
+class Function;
+
+/// Represents a particular input of a function.
+struct Edge {
+  explicit Edge(
+      const std::shared_ptr<Function>& function_ = nullptr,
+      uint32_t input_nr_ = 0)
```

Comment:
So I don't think that having a default value for `input_nr_` makes sense. It's not like 0 is a "fallback value". A better signature would be to have a no-argument constructor and one that takes both a function and an `input_nr`, to make sure you didn't forget to provide it

Reply:
Also, it's better to take the shared pointer as a value - otherwise you make an unnecessary copy e.g. in `gradient_edge()`. I'm also a bit concerned that this commit might increase the atomic instruction pressure. In general it would be better to have Variables keep an `Edge` and return a reference when asking for it. Otherwise if you do `if (var.grad_edge() != ...)` you're temporarily increfing a shared pointer, just to decref it right after the comparison

Reply:
I removed the default arguments. I think taking the shared pointer by value in the constructor would create a copy here:
```
if (const auto& gradient = grad_fn()) {
      return Edge(gradient, output_nr());
                           ^^^^^^^^
    }
```
while it would turn a copy into a move for:
```
return Edge(grad_accumulator(), 0);
                      ^^^^^^^^^^^^^^^^^^
```
so the choice is between (ref+copy, ref+copy) vs. (copy+move, move+move).
Maybe we could store the `grad_accumulator` inside `Variable` and have `grad_accumulator()` return a reference?

FYI, I just spent quite a decent amount of time trying to cache `gradient_edge` inside `VariableImpl` and only creating it the first time we call it and then returning a reference. However, there are too many places in the codebase where random code grabs the internals of `Variable` and assigns to the gradient function, breaking this caching. I believe `Variable` needs a proper API where you can `set_` things so that changes to internal state are encapsulated, allowing implementation changes like this, for example. I will leave it like this for now.

Reply:
Hmm we can't make `grad_accumulator()` return a (non-const) reference, because Variable only keeps a weakref to its grad accumulator (otherwise you end up with a reference cycle), so it needs to obtain a strong one every time you call `grad_accumulator()`. Yea, let's leave it